environs/simplestreams/datasource.go

```diff
@@ -73,26 +76,55 @@ type urlDataSource struct {
 	requireSigned        bool
 }
 
-// NewURLDataSource returns a new datasource reading from the specified baseURL.
-func NewURLDataSource(description, baseURL string, hostnameVerification utils.SSLHostnameVerification, priority int, requireSigned bool) DataSource {
-	return &urlDataSource{
-		description:          description,
-		baseURL:              baseURL,
-		hostnameVerification: hostnameVerification,
-		priority:             priority,
-		requireSigned:        requireSigned,
+// Config has values to be used in constructing a datasource.
+type Config struct {
+	// Description of the datasource
+	Description string
+
+	// BaseURL is the URL for this datasource.
+	BaseURL string
+
+	// HostnameVerification indicates whether to use self-signed credentials
+	// and not try to verify the hostname on the TLS/SSL certificates.
+	HostnameVerification utils.SSLHostnameVerification
+
+	// PublicSigningKey is the public key used to validate signed metadata.
+	PublicSigningKey string
+
+	// Priority is an importance factor for the datasource. Higher number means
+	// higher priority. This is will facilitate sorting data sources in order of
+	// importance.
+	Priority int
+
+	// RequireSigned indicates whether this datasource requires signed data.
+	RequireSigned bool
+}
+
+// Validate() checks that the baseURL is valid and the description is set.
+func (c *Config) Validate() error {
+	if c.Description == "" {
+		return errors.New("no description specified")
+	}
+	if _, err := url.Parse(c.BaseURL); err != nil {
+		return errors.Annotate(err, "base URL is not valid")
 	}
+	// TODO (hml) 2020-01-08
+	// Add validation for PublicSigningKey
+	return nil
 }
 
-// NewURLSignedDataSource returns a new datasource for signed metadata reading from the specified baseURL.
-func NewURLSignedDataSource(description, baseURL, publicKey string, hostnameVerification utils.SSLHostnameVerification, priority int, requireSigned bool) DataSource {
+// NewDataSource returns a new DataSource as defined
+// by the given config.
+func NewDataSource(cfg Config) DataSource {
+	// TODO (hml) 2020-01-08
+	// Move call to cfg.Validate() here and add return of error.
```

Comment:
Are you planning a different PR for this change? Given that this PR modifies all call-sites it might be worth making the change here.

Reply:
Agreed, unfortunately it's not a simple change.  Much of the codebase and tests expect for the urls etc not to validated and it's a longer PR I believe.  Got a few hours an and no end in site.  Since I'm already down the rabbit hole on this one, was going to do later.

Reply:
Fair enough